File: go/src/student_submissions/movielens.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

type recordWithIndex struct {
	record []string
	index  int
}

func main() {
	start := time.Now()
	err := SplitCsvFile("ratings.csv", "movies.csv", 10)
	if err != nil {
		fmt.Println("Error:", err)
	}
	duration := time.Since(start)
	fmt.Printf("Tiempo de ejecución: %v\n", duration)
}

// SplitCsvFile divide el archivo CSV en partes más pequeñas y cuenta los ratings por género
func SplitCsvFile(ratingsFile, moviesFile string, numFiles int) error {
	recordCh := make(chan recordWithIndex, 100)
	var wg sync.WaitGroup

	// Lanzamos goroutines para escribir en los archivos
	for i := 0; i < numFiles; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			outputFile := fmt.Sprintf("ratings_%d.csv", i+1)
			err := WriteCsvFile(outputFile, recordCh, numFiles, i)
			if err != nil {
				fmt.Printf("Error al escribir el archivo %s: %v\n", outputFile, err)
			}
		}(i)
	}

	// Leemos los registros en streaming y los enviamos al canal
	err := ReadRatingsCsvFile(ratingsFile, recordCh)
	if err != nil {
		return err
	}
	close(recordCh)
	wg.Wait()

	// Contamos los ratings por género y calculamos el promedio
	genresCount, genresAverage, err := CountRatingsByGenre(ratingsFile, moviesFile)
	if err != nil {
		return err
	}

	// Imprimimos los resultados
	for genre, count := range genresCount {
		avg := genresAverage[genre]
		fmt.Printf("Género: %s, Ratings: %d, Promedio: %.2f\n", genre, count, avg)
	}
	return nil
}

// ReadRatingsCsvFile lee el archivo CSV línea por línea y envía cada registro por el canal
func ReadRatingsCsvFile(filename string, recordCh chan<- recordWithIndex) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	reader := csv.NewReader(file)

	// Leemos la cabecera
	header, err := reader.Read()
	if err != nil {
		return err
	}
	// Enviamos la cabecera como el primer registro (índice -1) para que cada goroutine escriba la cabecera en su archivo
	recordCh <- recordWithIndex{record: header, index: -1}

	// Leemos y enviamos cada registro con su índice
	counter := 0
	for {
		record, err := reader.Read()
		if err != nil {
			break
		}
		recordCh <- recordWithIndex{record: record, index: counter}
		counter++
	}
	return nil
}

// WriteCsvFile escribe un archivo CSV con los registros asignados en función de un índice
func WriteCsvFile(filename string, recordCh <-chan recordWithIndex, numFiles, index int) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()

	for recordInfo := range recordCh {
		if recordInfo.index == -1 {
			if err := writer.Write(recordInfo.record); err != nil {
				return err
			}
			continue
		}
		if recordInfo.index%numFiles == index {
			if err := writer.Write(recordInfo.record); err != nil {
				return err
			}
		}
	}
	return nil
}

// CountRatingsByGenre cuenta el número de ratings y calcula el promedio por género
func CountRatingsByGenre(ratingsFile, moviesFile string) (map[string]int, map[string]float64, error) {
	genresCount := make(map[string]int)
	genresSum := make(map[string]float64)

	// Leemos el archivo de películas y construimos un mapa de géneros
	movies, err := os.Open(moviesFile)
	if err != nil {
		return genresCount, genresSum, err
	}
	defer movies.Close()
	moviesReader := csv.NewReader(movies)
	_, err = moviesReader.Read() // Saltamos la cabecera
	if err != nil {
		return genresCount, genresSum, err
	}
	movieGenres := make(map[string][]string)
	for {
		record, err := moviesReader.Read()
		if err != nil {
			break
		}
		movieID := record[0]
		genres := strings.Split(record[2], "|")
		movieGenres[movieID] = genres
	}

	// Leemos el archivo de ratings y contamos por género
	ratings, err := os.Open(ratingsFile)
	if err != nil {
		return genresCount, genresSum, err
	}
	defer ratings.Close()
	ratingsReader := csv.NewReader(ratings)
	_, err = ratingsReader.Read() // Saltamos la cabecera
	if err != nil {
		return genresCount, genresSum, err
	}
	for {
		record, err := ratingsReader.Read()
		if err != nil {
			break
		}
		movieID := record[1]
		rating := record[2]
		ratingFloat, err := strconv.ParseFloat(rating, 64)
		if err != nil {
			return genresCount, genresSum, err
		}
		genres, ok := movieGenres[movieID]
		if ok {
			for _, genre := range genres {
				genresCount[genre]++
				genresSum[genre] += ratingFloat
			}
		}
	}

	genresAverage := make(map[string]float64)
	for genre, sum := range genresSum {
		genresAverage[genre] = sum / float64(genresCount[genre])
	}

	return genresCount, genresAverage, nil
}

File: julia/src/student_submissions/movielens/movielens.jl
using CSV
using DataFrames
using Dates
using Base.Threads

function split_csv_file(ratings_file::String, movies_file::String, num_files::Int)
    # Leer los archivos CSV
    ratings = CSV.read(ratings_file, DataFrame)
    movies = CSV.read(movies_file, DataFrame)
    
    # Crear un canal para pasar registros
    record_channel = Channel{Tuple{Int, DataFrameRow}}(100)
    
    # Definir las tareas para escribir en los archivos usando hilos
    tasks = []
    for i in 1:num_files
        push!(tasks, @spawn write_csv_file("ratings_$(i).csv", record_channel, num_files, i))
    end
    
    # Leer y enviar los registros al canal
    read_ratings_csv_file(ratings, record_channel, num_files)
    close(record_channel)
    
    # Esperar a que todas las tareas terminen
    for task in tasks
        wait(task)
    end
    
    # Contar los ratings por género y calcular el promedio
    genres_count, genres_avg = count_ratings_by_genre(ratings, movies)
    
    # Imprimir los resultados
    for (genre, count) in genres_count
        avg = genres_avg[genre]
        println("Género: $genre, Ratings: $count, Promedio: $avg")
    end
    
    # Guardar los resultados en un archivo CSV
    write_genre_summary_csv("ratings_summary.csv", genres_count, genres_avg)
end

function read_ratings_csv_file(ratings::DataFrame, record_channel::Channel{Tuple{Int, DataFrameRow}}, num_files::Int)
    counter = 0
    for row in eachrow(ratings)
        put!(record_channel, (counter % num_files, row))
        counter += 1
    end
end

function write_csv_file(filename::String, record_channel::Channel{Tuple{Int, DataFrameRow}}, num_files::Int, index::Int)
    rows_to_write = DataFrame()
    
    # Acumular las filas que corresponden a este archivo
    for (i, row) in record_channel
        if i == index - 1
            push!(rows_to_write, row)
        end
    end
    
    # Escribir todas las filas al archivo de una vez
    CSV.write(filename, rows_to_write)
end

function count_ratings_by_genre(ratings::DataFrame, movies::DataFrame)
    genres_count = Dict{String, Int}()
    genres_sum = Dict{String, Float64}()
    
    # Crear un mapa de géneros de las películas
    movie_genres = Dict{String, Vector{String}}()
    for row in eachrow(movies)
        movie_id = string(row[:movieId])
        genres = split(row[:genres], "|")
        movie_genres[movie_id] = genres
    end
    
    # Contar los ratings por género
    for row in eachrow(ratings)
        movie_id = string(row[:movieId])
        rating = row[:rating]
        if haskey(movie_genres, movie_id)
            genres = movie_genres[movie_id]
            for genre in genres
                genres_count[genre] = get(genres_count, genre, 0) + 1
                genres_sum[genre] = get(genres_sum, genre, 0.0) + rating
            end
        end
    end
    
    # Calcular el promedio de ratings por género
    genres_avg = Dict{String, Float64}()
    for (genre, sum) in genres_sum
        genres_avg[genre] = sum / genres_count[genre]
    end
    
    return genres_count, genres_avg
end

function write_genre_summary_csv(filename::String, genres_count::Dict{String, Int}, genres_avg::Dict{String, Float64})
    # Crear un DataFrame con los resultados
    summary_df = DataFrame(genre = String[], ratings_count = Int[], average_rating = Float64[])
    
    for (genre, count) in genres_count
        avg = genres_avg[genre]
        push!(summary_df, (genre, count, avg))
    end
    
    # Escribir el DataFrame en un archivo CSV
    CSV.write(filename, summary_df)
    println("Resumen de ratings por género guardado en $filename")
end

# Ejemplo de uso
start_time = now()
split_csv_file("ratings.csv", "movies.csv", 10)
elapsed_time = now() - start_time
println("Tiempo de ejecución: $elapsed_time")


